README.md
# pagespeedParseR  v1.0.0
![Lifecycle_Status](https://img.shields.io/badge/lifecycle-maturing-blue.svg)
[![Build status](https://travis-ci.org/Leszek-Sieminski/pagespeedParseR.svg?branch=master)](https://travis-ci.org/Leszek-Sieminski/pagespeedParseR)
[![Coverage status](https://codecov.io/gh/Leszek-Sieminski/pagespeedParseR/branch/master/graph/badge.svg)](https://codecov.io/github/Leszek-Sieminski/pagespeedParseR?branch=master)

R wrapper for Google Pagespeed Insights API

* [News](#news)
* [What is Google Pagespeed Insights?](#what-is-google-pagespeed-insights)
* [Other Pagespeed API packages in R](#other-pagespeed-packages-in-r)
* [Why pagespeedParseR when there are other packages?](#but-why-another-r-package-for-pagespeed-api)
* [Features](#features)
* [Acquiring API access token](#acquiring-api-access-token)
* [Installation](#installation)
* [Authentication](#authentication)
* [Usage](#usage)

## News 

##### **2019-07-...**, v1.0.0 (package api shift + release):
Big paradigm shift, may break exisiting code!

* there is only **one** function that downloads just the nested lists - *download_lighthouse()*. Parameters *output_type* & *long_result* are now obsolete (and will break the function) and were moved into specialised extraction functions (see below)
* to extract the data from the export in the form of data frames, use new functions: *extract_lighthouse_...()*. You can choose to export 1) just the category scores, 2) most of the data from all audits or 3) detailed data from choosen audit (including errors & recommendations)
* pagespeedParseR should be able to handle thousands of URLs thanks to the new cache functionality! The results are cached into temporary binary file 'db.llo' saved in a current working directory and thus should not cause memory allocation issues
* you can choose to export the API output as a binary file to location of your choosing and access its data without loading it up into the global enviroment. You can just use the pointer overloading. See more in *download_lighthouse()* documentation
* if you happen to find a bug, don't hesitate to inform me and/or [add an issue](https://github.com/Leszek-Sieminski/pagespeedParseR/issues)

To find out more, please read the documentation.

## What is Google Pagespeed Insights?
Google Pagespeed is an online tool that identifies performance issues for a given URL and provides suggestions and optimizations. See [more details](https://developers.google.com/speed/pagespeed/insights/?hl=pl).

Google Lighthouse is another tool for webdevelopers that helps in fixing page performance. It was recently featured in Pagespeed API (version 5). You can try it via Chrome browser, see [more details](https://developers.google.com/web/tools/lighthouse/).

## Other Pagespeed packages in R
There are other R packages for Pagespeed that I find very valuable:

* googlePageSpeedR - [Phippsy/googlePageSpeedR](https://github.com/Phippsy/googlePageSpeedR)
* gpagespeed - [simitpatel/gpagespeed](https://github.com/simitpatel/gpagespeed)
* pagespeed - [mhairi/pagespeed](https://github.com/mhairi/pagespeed)

## But why another R package for Pagespeed API?
Parsing **API response returns nested list object** in R which isn't very convenient to extract data from (at least in R), so the outputs of above packages' functions are limited (doesn't contain recommendations). The goal of pagespeedParseR is to **create simple output in form of data frame** which contains both performance scores and most of recommendations + list of errors. Alternatively, **user can decide to obtain the output in raw form of nested list** and parse it by himself/herself if that's not enough.

## Features
* Authenticate with an API key
* Choose report content: use download_pagespeed() for classic PageSpeed v4 results OR download_lighthouse() for new Lighthouse (PageSpeed v5) results
* Choose output format: simple (conveniently parsed data frame with most important info) or raw (nested list with all the data provided by the API)
* Query multiple URLs in one function call and keep information which URL's failed the check
* Query every page for both Desktop and Mobile in one function call
* Control API limits usage with simple interval mechanism

## Acquiring API access token
Visit your [Google Developers Console](https://console.developers.google.com/) page, create the project, [switch on](https://console.developers.google.com/apis/library/pagespeedonline.googleapis.com) Pagespeed API and copy your API access token into R from the credentials screen.

## Installation

```r
install.packages("devtools")
devtools::install_github("Leszek-Sieminski/pagespeedParseR")
```

## Authentication

```r
library(pagespeedParseR)
api_key <- "12345" # example
auth_pagespeed(api_key)
```

## Usage

### Download & load

```r
# libraries -------------------------------------------------------------------
install.packages("devtools")
library(devtools)
devtools::install_github("Leszek-Sieminski/pagespeedParseR")
library(pagespeedParseR)
```

### Authentication
```r
# authentication --------------------------------------------------------------
auth_pagespeed("12345") # Not run, example
```

### Downloading Lighthouse reports as objects to global enviroment
```r
# downloading Lighthouse as objects -------------------------------------------
obj1 <- download_lighthouse(
  url        = c("https://www.w3.org/"), # single URL
  categories = c("performance",
                 "accessibility",
                 "best-practices",
                 "pwa",
                 "seo"))

obj2 <- download_lighthouse(
  url        = c("https://www.w3.org/",     # multi-URL
                 "https://www.google.com"),
  categories = c("performance",
                 "accessibility",
                 "best-practices",
                 "pwa",
                 "seo"))
```

### Downloading Lighthouse as references to physical cache files on disk
```r
# downloading Lighthouse as references to physical files on disk --------------
ref1 <- download_lighthouse(
  url            = c("https://www.w3.org/"), # single URL
  categories     = c("performance",
                     "accessibility",
                     "best-practices",
                     "pwa",
                     "seo"),
  as_reference   = TRUE,
  reference_path = "ref_path_1.llo")

ref2 <- download_lighthouse(
  url            = c("https://www.w3.org/",     # multi-URL
                     "https://www.google.com"),
  categories     = c("performance",
                 "accessibility",
                 "best-practices",
                 "pwa",
                 "seo"),
  as_reference   = TRUE,
  reference_path = "ref_path_2.llo")
```

### Extracting only Lighthouse category scores
```r
# extracting only category scores ---------------------------------------------
scor_ref_1 <- pagespeedParseR::extract_lighthouse_scores(ref1)
scor_ref_2 <- pagespeedParseR::extract_lighthouse_scores(ref2)
scor_obj_1 <- pagespeedParseR::extract_lighthouse_scores(obj1)
scor_obj_2 <- pagespeedParseR::extract_lighthouse_scores(obj2)

str(scor_obj_2)
# 'data.frame':	2 obs. of  7 variables:
# $ url           : chr  "https://www.w3.org/" "https://www.google.com/"
# $ device        : chr  "desktop" "desktop"
# $ performance   : num  1 1
# $ accessibility : num  0.78 0.89
# $ best.practices: num  0.77 0.92
# $ seo           : num  1 0.8
# $ pwa           : num  0.54 0.46
```

### Extracting most important Lighthouse data from all audits
```r
# extracting most important data from all audits ------------------------------
gen_ref_1  <- pagespeedParseR::extract_lighthouse_general_info(ref1)
gen_obj_1  <- pagespeedParseR::extract_lighthouse_general_info(obj1)
gen_obj_2  <- pagespeedParseR::extract_lighthouse_general_info(obj2)
gen_ref_2  <- pagespeedParseR::extract_lighthouse_general_info(ref2)

str(gen_ref_2)
# 'data.frame':	2 obs. of  1630 variables:
# $ device               : chr  "desktop" "desktop"
# $ url                  : chr  "https://www.w3.org/" "https://www.google.com/"
# $ finalUrl             : chr  "https://www.w3.org/" "https://www.google.com/"
# $ status_code          : num  200 200
# $ score.accessibility  : num  0.78 0.89
# $ score.best-practices : num  0.77 0.92
# $ score.performance    : num  1 1
# $ score.pwa            : num  0.54 0.46
# $ score.seo            : num  1 0.8
# ... (truncated)
```

### PageSpeed reports as wide data frames (most important data)
```r
# PageSpeed reports -  Data Frames --------------------------------------------
# download simple data frame with Pagespeed report for Google.com
ps_df_1 <- download_pagespeed(url = "https://www.google.com", output_type = "simple")

# run Pagespeed reports for Google.com & Bing.com for mobile and
# return in a data frame with most important columns
ps_df_2 <- download_pagespeed(url = c("https://www.google.com", 
                                      "https://www.bing.com/"), 
                              output_type = "simple",      # return the results in a wide data frame
                              strategy = "mobile",         # run tests for mobile
                              interval = 1)                # wait 1 second between the calls to API 

# run Pagespeed reports for Google.com & Bing.com for both desktop & mobile and
# return in a data frame with most important columns                              
ps_df_3 <- download_pagespeed(url = c("https://www.google.com", 
                                      "https://www.bing.com/"), 
                              output_type = "simple",      # return the results in a wide data frame
                              strategy = c("desktop",      # check both desktop and mobile, bind
                                           "mobile"), 
                              interval = 2)                # wait 2 seconds between the calls to API 
```

### PageSpeed reports as nested lists (all data)
```r
# PageSpeed reports - Nested Lists --------------------------------------------
# download nested list with Pagespeed report for Google.com
ps_nl_1 <- download_pagespeed(url = "https://www.google.com", 
                              output_type = "raw")

# run Pagespeed for Google.com & Bing.com for desktop and
# return in a nested list with all possible data
ps_nl_2 <- download_pagespeed(url = c("https://www.google.com", 
                                      "https://www.bing.com/"), 
                              output_type = "raw", 
                              strategy = "desktop", 
                              interval = 1)

# check "Performance" for Google.com & Bing.com for both desktop & mobile and
# return in a nested list with all possible data
ps_nl_3 <- download_pagespeed(url = c("https://www.google.com", 
                                      "https://www.bing.com/"), 
                              output_type = "raw", 
                              strategy = c("desktop", 
                                           "mobile"), 
                              interval = 2)
```


NEWS.md
## Plans for future:
* adding parallelization (multicore) in extraction on Windows & Linux
* speeding things up
* release on CRAN

## News
##### **2019-07-...**, v1.0.0 (package api shift + release):
* there is only **one** function that downloads just the nested lists - *download_lighthouse()*. Parameters *output_type* & *long_result* are now obsolete (and will break the function) and were moved into specialised extraction functions (see below)
* to extract the data from the export in the form of data frames, use new functions: *extract_lighthouse_...()*. You can choose to export just the category scores, most of the data from all audits or detailed data from choosen audit (including errors & recommendations)
* pagespeedParseR should be able to handle thousands of URLs thanks to the new cache functionality! The results are cached into temporary binary file 'db.llo' saved in a current directory and should not cause memory allocation issues
* you can choose to export the API output as a binary file to location of your choosing and can access its data without loading it up into the global enviroment. You can just use the pointer overloading. See more in *download_lighthouse()* documentation
* if you happen to find a bug, don't hesitate to inform me and/or add an [Issue](https://github.com/Leszek-Sieminski/pagespeedParseR/issues)

##### **2019-07-10**, v0.3.1.9000 (bugfixes):
* small bugfixes to *long_result = T* parameter in *download_lighthouse(..., output_type = "simple")*
* small bugfixes to tests

##### **2019-07-05**, v0.3.0.9000 (Lighthouse overhaul):
* big overhaul of *download_lighthouse()* function. Parsing to data frame with *output_type = "simple"* parameter will now provide much more data. However, **it can generate literally hundreds/thousands of columns** (up to ~2500). What is more, the **number of columns IS NOT STABLE**, because it depends on the number of found errors and/or their type
* to ease the pain of dealing with such data frames, I added *long_result* parameter that defaults to *FALSE*. Setting it to TRUE will force the function to spread the data frame into messy, long-like form that I hope to be easier to comprehend
* the behaviour of download_lighthouse(output_type = "raw") or other functions didn't change



R/lh_simple_1_s.R
#' Download Pagespeed v4 raport for an URL as a data frame
#'
#' @param url string. The URL to fetch and analyze
#' @param key string. Pagespeed API key to authenticate. Defaults to
#'     "PAGESPEED_API_KEY" enviroment variable.
#' @param strategy string. The analysis strategy to use. Options: "desktop" or
#'     "mobile". Defaults to "desktop"
#' @param categories string. A Lighthouse category/categories to run.
#'     Defaults to "performance". See more in Details section
#' @param interval numeric. Number of seconds to wait between multiple queries.
#'     Defaults to 0.5 second.
#' @param locale string. The locale used to localize formatted results
#' @param utm_campaign string. Campaign name for analytics. Defaults to NULL
#' @param utm_source string. Campaign source for analytics. Defaults to NULL
#'
#' @details The \code{category} parameter regulates which of the tests'
#'     categories from Lighthouse are to be run. You can select more than
#'     one.
#'     Options: "accessibility", "best-practices", "performance", "pwa",
#'     "seo".
#'
#' @return data frame
#'
#' @importFrom httr GET
#' @importFrom httr http_type
#' @importFrom httr content
#' @importFrom jsonlite fromJSON
#'
#' @examples
#' \dontrun{
#' single_url_simple_output_5 <- lh_simple_1("https://www.google.com/")
#' }
lh_simple_1 <- function(url, key = Sys.getenv("PAGESPEED_API_KEY"),
                        strategy = NULL, categories = "performance",
                        interval = 0.5,
                        locale = NULL,
                        utm_campaign = NULL, utm_source = NULL)
{
  # safety net ----------------------------------------------------------------
  if (is.null(key) | nchar(key) == 0){stop("API key is a NULL or has length = 0. Please check it and provide a proper API key.", call. = FALSE)}

  assert_that(not_empty(url), is.string(url) & length(url) > 0, grepl(".", url, fixed = T),
              is.string(key), is.null(strategy) ||
                (is.character(strategy) & strategy %in% c("desktop", "mobile")),
              is.null(categories) ||
                (is.character(categories) & categories %in%
                   c("accessibility", "best-practices", "performance", "pwa", "seo")),
              is.number(interval) & interval >= 0 & interval <= 120,
              (is.string(locale) & nchar(locale) > 0) || is.null(locale),
              is.string(utm_campaign) | is.null(utm_campaign),
              is.string(utm_source)   | is.null(utm_source))

  # downloading ---------------------------------------------------------------
  req <- GET(
    url = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed",
    query = list(
      url = url, strategy = strategy, key = key, locale = locale,
      category = categories[1],
      category = if (length(categories) >= 2) {categories[2]} else {NULL},
      category = if (length(categories) >= 3) {categories[3]} else {NULL},
      category = if (length(categories) >= 4) {categories[4]} else {NULL},
      category = if (length(categories) == 5) {categories[5]} else {NULL},
      utm_campaign = utm_campaign,
      utm_source = utm_source))

  # parsing -------------------------------------------------------------------
  # httr::stop_for_status(req) # we don't want to stop for error
  # as we want to know which URL's wasn't properly returned
  if (req$status_code == 200) {
    # 01 json check --------------------------------------------------------------
    if (http_type(req) != "application/json") {stop("API did not return json", call. = FALSE)}

    # 02 extracting from JSON -------------------------------------------------
    parsed  <- fromJSON(content(req, as = "text", type = "application/json", encoding = "UTF-8"))

    # 03 creating baseline data frame -----------------------------------------
    baseline <- data.frame(
      device           = `if`(is.null(strategy), "desktop", strategy),
      url              = url, #`if`(is.null(parsed$loadingExperience$initial_url), NA, parsed$loadingExperience$initial_url),
      status_code      = req$status_code,
      stringsAsFactors = FALSE)

    # 04 intermediary audit object -----------------------------------------------
    audits <- parsed$lighthouseResult$audits

    # 05 report category object --------------------------------------------------
    if ("performance" %in% categories) {
      pm <- data.frame(
        category = "performance",
        report_name = gsub("-", "_", parsed$lighthouseResult$categories$performance$auditRefs$id,
                           fixed = TRUE),
        stringsAsFactors = FALSE)
    }

    if ("accessibility" %in% categories) {
      acc <- data.frame(
        category = "accessibility",
        report_name = gsub("-", "_", parsed$lighthouseResult$categories$accessibility$auditRefs$id,
                           fixed = TRUE),
        stringsAsFactors = FALSE)
    }

    if ("best-practices" %in% categories) {
      bp <- data.frame(
        category = "best-practices",
        report_name = gsub("-", "_", parsed$lighthouseResult$categories$`best-practices`$auditRefs$id,
                           fixed = TRUE),
        stringsAsFactors = FALSE)
    }

    if ("pwa" %in% categories) {
      pwa <- data.frame(
        category = "pwa",
        report_name = gsub("-", "_", parsed$lighthouseResult$categories$pwa$auditRefs$id,
                           fixed = TRUE),
        stringsAsFactors = FALSE)
    }

    if ("seo" %in% categories) {
      seo <- data.frame(
        category = "seo",
        report_name = gsub("-", "_", parsed$lighthouseResult$categories$seo$auditRefs$id,
                           fixed = TRUE),
        stringsAsFactors = FALSE)
    }

    report_cat_df <- rbind(
      if ("performance" %in% categories) {pm},
      if ("accessibility" %in% categories) {acc},
      if ("best-practices" %in% categories) {bp},
      if ("pwa" %in% categories) {pwa},
      if ("seo" %in% categories) {seo}
    )

    # 06 basic lighthouse data extraction -------------------------------------
    # basic <- fun_lh_basic_extract(audits, report_cat_df)
    # basic <- pagespeedParseR:::fun_lh_basic_extract(audits, report_cat_df)
    results_load_exp <- fun_lh_enhanced_extract(parsed$loadingExperience)
    results_audits   <- fun_lh_enhanced_extract(audits, report_cat_df)
    full_results     <- cbind(baseline, results_load_exp, results_audits)

    # 07 missing columns in case of mobile ------------------------------------
    # if ((grepl("desktop", strategy) || is.null(strategy)) & "performance" %in% categories) {
    #   full_results$performance_first_contentful_paint_3g_description   <- NA
    #   full_results$performance_first_contentful_paint_3g_score         <- NA
    #   full_results$performance_first_contentful_paint_3g_display_value <- NA
    # }

    if ("performance" %in% categories) {full_results$score.performance <- parsed$lighthouseResult$categories$performance$score}
    if ("accessibility" %in% categories) {full_results$score.accessibility <- parsed$lighthouseResult$categories$accessibility$score}
    if ("best-practices" %in% categories) {full_results$score.best_practices <- parsed$lighthouseResult$categories$`best-practices`$score}
    if ("pwa" %in% categories) {full_results$score.pwa <- parsed$lighthouseResult$categories$pwa$score}
    if ("seo" %in% categories) {full_results$score.seo <- parsed$lighthouseResult$categories$seo$score}

    # 08 sorting the columns --------------------------------------------------
    colnames_v <- colnames(full_results) # capturing column names
    colnames_sorted <- sort(colnames_v) # sorting column names alphabetically

    # deleting baseline columns (they need to go first!)
    col_scores <- colnames_sorted[grepl("^score.*", colnames_sorted)] # main scores
    colnames_sorted <- colnames_sorted[!colnames_sorted %in% c('device', 'url', 'status_code', col_scores)]
    # adding baseline columns at the beggining of the df
    colnames_sorted <- c(c('device', 'url', 'status_code', col_scores), colnames_sorted)
    full_results <- full_results[, colnames_sorted] # choosing the df columns in order

    # full_results_tf1 <- tidyr::gather(full_results, parameter, value, -url)
    # full_results_tf2 <- tidyr::gather(full_results, parameter, value)
    # full_results_tf <- tidyr::gather(full_results, parameter, value, -url)
    # colnames(full_results_tf)[3] <- full_results_tf$url[1]
    # full_results_tf$url <- NULL

    # 10 returning ------------------------------------------------------------
    return(full_results)
  } else {
    # else NA df --------------------------------------------------------------
    # if there were no results, create placeholder to keep track which URL failed
    parsed  <- fromJSON(content(req, as = "text", type = "application/json", encoding = "UTF-8"))
    full_results <- data.frame(
      device           = `if`(is.null(strategy), "desktop", strategy),
      url              = url, #`if`(is.null(parsed$loadingExperience$initial_url), NA, parsed$loadingExperience$initial_url),
      status_code      = req$status_code,
      stringsAsFactors = FALSE)

    full_results$url <- as.character(full_results$url)
    Sys.sleep(0.5 + interval) # optional waiting to keep API limits happy
    return(full_results)
  }
}
